src/nnet3/nnet-simple-component.h

```diff
@@ -1643,6 +1643,109 @@ class MaxpoolingComponent: public Component {
 };
 
 /**
+ * Maxpooling3dComponent implements 3d-maxpooling.
+ * This component is designed to be used after a ConvolutionComponent
+ * so that the input matrix is propagated from a 2d/3d convolutional layer.
+ * This component supports performing max pooling along the three axes.
+ * Pooling with overlapping on each axis is supported.
+ * Input : A matrix where each row is a vectorized 3D-tensor.
+ *        The 3D tensor has dimensions
+ *        x: (e.g. time)
+ *        y: (e.g. frequency)
+ *        z: (e.g. channels like number of filters in the ConvolutionComponent)
+ *
+ *        The component only supports input vectorizations of type zyx
+ *        which is the default output vectorization type of a ConvolutionComponent.
+ *        For the description of vectorized 3d tensor, please read the comment
+ *        at the ConvolutionComponent.
+ *
+ * Output : The output is also a 3D tensor vectorized in the zyx format.
+ *
+ * For information on the hyperparameters and parameters of this component see
+ * the variable declarations.
+ *
+ *
+ */
+
+class Maxpooling3dComponent: public Component {
+ public:
+  explicit Maxpooling3dComponent(int32 input_x_dim, int32 input_y_dim, int32 input_z_dim,
+                                 int32 pool_x_size, int32 pool_y_size, int32 pool_z_size,
+                                 int32 pool_x_step, int32 pool_y_step, int32 pool_z_step
+                                 ) {
+    Init(input_x_dim, input_y_dim, input_z_dim,
+         pool_x_size, pool_y_size, pool_z_size,
+         pool_x_step, pool_y_step, pool_z_step);
+  }
+  Maxpooling3dComponent(): input_x_dim_(0), input_y_dim_(0), input_z_dim_(0),
+                           pool_x_size_(0), pool_y_size_(0), pool_z_size_(0),
+                           pool_x_step_(0), pool_y_step_(0), pool_z_step_(0) { }
+  virtual int32 InputDim() const;
+  virtual int32 OutputDim() const;
+
+  virtual std::string Info() const;
+  virtual void InitFromConfig(ConfigLine *cfl);
+  virtual std::string Type() const { return "Maxpooling3dComponent"; }
+  virtual int32 Properties() const {
+    return kSimpleComponent|kBackpropNeedsInput|kBackpropNeedsOutput|
+	    kBackpropAdds;
```

Comment:
Vigay: regarding when to set kBackpropAdds and kPropagateAdds- it's based
on what's most convenient and efficient in the backprop and propagate
code.  For instance, if the propagate code uses AddMatMat, you should set
kPropagateAdds because you'd need a separate statement to zero it (and
there might be situations where you are actually adding to something).
Prefer addition in situations where you would otherwise have to zero the
matrix.
Dan

On Sat, Jan 23, 2016 at 7:45 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-simple-component.h
> <https://github.com/kaldi-asr/kaldi/pull/460#discussion_r50628209>:
>
> > +    Init(input_x_dim, input_y_dim, input_z_dim,
> > +         pool_x_size, pool_y_size, pool_z_size,
> > +         pool_x_step, pool_y_step, pool_z_step);
> > +  }
> > +  Maxpooling3dComponent(): input_x_dim_(0), input_y_dim_(0), input_z_dim_(0),
> > +                           pool_x_size_(0), pool_y_size_(0), pool_z_size_(0),
> > +                           pool_x_step_(0), pool_y_step_(0), pool_z_step_(0) { }
> > +  virtual int32 InputDim() const;
> > +  virtual int32 OutputDim() const;
> > +
> > +  virtual std::string Info() const;
> > +  virtual void InitFromConfig(ConfigLine *cfl);
> > +  virtual std::string Type() const { return "Maxpooling3dComponent"; }
> > +  virtual int32 Properties() const {
> > +    return kSimpleComponent|kBackpropNeedsInput|kBackpropNeedsOutput|
> > +	    kBackpropAdds;
>
> @danpovey <https://github.com/danpovey> it is not clear to me when
> kBackpropAdds and kPropagateAdds bits have to be set. Is it sufficient to
> choose some value for these bits and honor that in the prop and backprop
> codes or should the decision of setting these bits be based on other
> information ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/460/files#r50628209>.
>